es-core/src/components/ImageGridComponent.h

```diff
@@ -90,27 +140,246 @@ class ImageGridComponent : public IList<ImageGridData, T>
+
+	CursorRange mCursorRange;	
+	int mCurrentLoad = 0;					// The current loaded in texture
+	bool bLoading = false;					// Loading in textures in the cursor range.
+	bool bUnloaded = false;					// No longer loading and just finished unloading old textures.
+
+	int mPrevIndex = 0;
+	int mPrevScrollTier = 0;
+	int mCurrentDirection = MOVING_DOWN;	// Which direction the user is moving (Orientation based on grid, not index)
+
+	std::vector< std::shared_ptr<GridTileComponent> > mTiles;
+
+	std::shared_ptr<ThemeData> mTheme;
+	bool bThemeLoaded = false;
+	bool bVisible = false;
 };
 
 template<typename T>
-ImageGridComponent<T>::ImageGridComponent(Window* window) : IList<ImageGridData, T>(window)
+ImageGridComponent<T>::ImageGridComponent(Window* window, int modGridSize) : IList<ImageGridData, T>(window)
 {
 	mEntriesDirty = true;
+	setMargin(Eigen::Vector2f(24, 24));
+	mDesiredGridSize.x() = 0;
+	mDesiredGridSize.y() = 0;
+	mMissingBoxartTexture = TextureResource::get(":/blank_game.png");
+}
+
+template<typename T>
+ImageGridComponent<T>::~ImageGridComponent() {
+	mTiles.clear();
+}
+
+template<typename T>
+void ImageGridComponent<T>::clear(bool clearall) {
+	unloadTextures(clearall);
+	mEntriesDirty = true;
+	mCurrentDirection = MOVING_DOWN;
+
+	mPrevIndex = 0;
+	bLoading = false;
+	bUnloaded = false;
+	mCurrentLoad = 0;
+
+	IList<ImageGridData, T>::clear();
+	mTiles.clear();
+}
+
+template<typename T>
+int ImageGridComponent<T>::getEntryCount() {
+	return mEntries.size();
+}
+
+template<typename T>
+int ImageGridComponent<T>::getCursorIndex() {
+	return static_cast<IList< ImageGridData, T >*>(this)->getCursorIndex();
+}
+
+template<typename T>
+void ImageGridComponent<T>::remove() {
+	static_cast<IList< ImageGridData, T >*>(this)->pop_back();
+
+	mEntriesDirty = true;
 }
 
 template<typename T>
-void ImageGridComponent<T>::add(const std::string& name, const std::string& imagePath, const T& obj)
+void ImageGridComponent<T>::add(const std::string& name, const std::string& imagePath, const T& obj, bool loadTextureNow)
 {
 	typename IList<ImageGridData, T>::Entry entry;
 	entry.name = name;
 	entry.object = obj;
-	entry.data.texture = ResourceManager::getInstance()->fileExists(imagePath) ? TextureResource::get(imagePath) : TextureResource::get(":/button.png");
+	entry.strdata = imagePath;
+	if (obj->getType() == 2) entry.data.texture = TextureResource::get(":/folder.png");
+	else {
+		if (loadTextureNow) entry.data.texture = ResourceManager::getInstance()->fileExists(imagePath) ? TextureResource::get(imagePath) : TextureResource::get(":/blank_game.png");
+		else entry.data.texture = TextureResource::get(":/frame.png");
+	}
+
 	static_cast<IList< ImageGridData, T >*>(this)->add(entry);
 	mEntriesDirty = true;
 }
 
 template<typename T>
+FileData* ImageGridComponent<T>::getSelectedObject() {
+	return mEntries.at(getCursorIndex()).object;
+}
+
+template<typename T>
+void ImageGridComponent<T>::unloadTextures(bool unloadAll) {
```

Comment:
maybe make this method save / cheap to call when empty?

Reply:
What do you mean ?

Reply:
hmm, I don't remember :-/
I think my reasoning was something like, we call this from several locations, but we do not want to include calling-criteria at every calling-instance, so maybe just always call, even if the list is empty to begin with. But there is nothing to really optimize here anyways, so...
Never mind!